### PR TITLE
Single pref for wallet mnemonic

### PR DESCRIPTION
**Pull request by supermassive**
Resolves https://github.com/brave/brave-browser/issues/38011

A set of refactoring which aim to reduce complexity of keyrings/encryptors code:
- There is a single pref for mnemonic
- There is a single pref for encryptor's salt
- There is only 'encryptor_' instance now
- Each encrypted piece of data(mnemonic and all imported private keys) is stored as a dict of ciphertext and random nonce
- Prefs upgrade is done when valid password is provided by user
- Wallet restore is refactored to be either a wallet unlock or new wallet creation with given mnemonic. RestoreKeyring is gone.

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Review by mkarolin — APPROVED** [review]
`strings`++

File: ios/brave-ios/Sources/BraveWallet/Preview Content/MockKeyringService.swift [hunk 1]
```diff
@@ -118,7 +118,7 @@ class MockKeyringService: BraveWalletKeyringService {
     completion(isBackedUp)
   }
 
-  func mnemonicForDefaultKeyring(password: String, completion: @escaping (String) -> Void) {
+  func walletMnemonic(password: String, completion: @escaping (String) -> Void) {
```
Comment (StephenHeaps) [inline ios/brave-ios/Sources/BraveWallet/Preview Content/MockKeyringService.swift:121]: Need optional string in completion closure now, same in `createWallet` function.
```suggestion
  func walletMnemonic(password: String, completion: @escaping (String?) -> Void) {
```

File: ios/brave-ios/Sources/BraveWallet/Crypto/Stores/KeyringStore.swift
Comment (StephenHeaps) [inline ios/brave-ios/Sources/BraveWallet/Crypto/Stores/KeyringStore.swift:1]: A couple small compilation fixes needed here. 
In `func createWallet(password: String, completion: ((String?) -> Void)? = nil)`:
`if !mnemonic.isEmpty {` -> `if mnemonic != nil {`

In `func recoveryPhrase(password: String, completion: @escaping ([RecoveryWord]) -> Void) {`, ned to update function name + unwrap `phrase` in completion, ex:
```
  func recoveryPhrase(password: String, completion: @escaping ([RecoveryWord]) -> Void) {
    keyringService.walletMnemonic(password: password) { phrase in
      guard let phrase else {
        completion([])
        return
      }
      let words =
        phrase
        .split(separator: " ")
        .enumerated()
        .map { RecoveryWord(value: String($0.element), index: $0.offset) }
      completion(words)
    }
  }
```

File: components/brave_wallet_ui/common/slices/endpoints/wallet.endpoints.ts [hunk 2]
```diff
@@ -131,6 +131,11 @@ export const walletEndpoints = ({
           const { keyringService } = api
 
           const result = await keyringService.createWallet(arg.password)
+          if (!result.mnemonic) {
+            return {
+              error: 'Unable to create wallet'
+            }
```
Comment (josheleonard) [inline components/brave_wallet_ui/common/slices/endpoints/wallet.endpoints.ts:137]: let's throw an error and let the catch block return the error message

**Review by josheleonard — APPROVED** [review]
Approved with nit

**Review by StephenHeaps — APPROVED** [review]
iOS++

**Review by darkdh — APPROVED** [review]
Changes look good! Codes in KeyringService is extremely organized now, especially those functions around restore wallet. 
Just a few nits.

File: ios/brave-ios/Sources/BraveWallet/Crypto/Onboarding/RestoreWalletView.swift [hunk 3]
```diff
@@ -271,7 +271,7 @@ struct RestoreWalletView: View {
     keyringStore.restoreWallet(
       words: recoveryWords,
       password: password,
-      isLegacyBraveWallet: recoveryWords.count == .legacyWalletRecoveryPhraseNumber
+      isLegacyEthSeedFormat: recoveryWords.count == .legacyWalletRecoveryPhraseNumber
```
Comment (supermassive) [inline ios/brave-ios/Sources/BraveWallet/Crypto/Onboarding/RestoreWalletView.swift:274]: @StephenHeaps it seems like a bug to me:
For desktop `isLegacyEthSeedFormat` is set to true when there are 24 words and a checkbox is toggled, but for ios there is no checkbox
https://github.com/brave/brave-core/blob/master/components/brave_wallet_ui/page/screens/restore-wallet/restore-wallet.tsx#L206-L217

Comment (StephenHeaps) [reply ios/brave-ios/Sources/BraveWallet/Crypto/Onboarding/RestoreWalletView.swift:274]: @supermassive iOS has a button to toggle between 12 / 24 word fields showing:
https://github.com/brave/brave-core/blob/master/ios/brave-ios/Sources/BraveWallet/Crypto/Onboarding/RestoreWalletView.swift#L152-L174

12-word | 24-word
--|--

File: components/brave_wallet/browser/keyring_service.cc [hunk 4]
```diff
@@ -2397,43 +2370,24 @@ void KeyringService::IsStrongPassword(const std::string& password,
...
 
-  auto mnemonic = encryptor->Decrypt(*encrypted_mnemonic, *nonce);
-  return mnemonic && !mnemonic->empty();
+  return DecryptWalletMnemonicFromPrefs(profile_prefs_, *encryptor);
 }
 
 void KeyringService::ValidatePassword(const std::string& password,
                                       ValidatePasswordCallback callback) {
+  MaybeMigratePBKDF2Iterations(password);
+  MaybeMigrateToWalletMnemonic(password);
```
[the remark this answers is not in the document]
Comment (darkdh) [reply components/brave_wallet/browser/keyring_service.cc]: sounds good.

File: ios/brave-ios/Sources/BraveWallet/Crypto/Onboarding/RestoreWalletView.swift (hunk 3 above)
Comment (supermassive) [reply ios/brave-ios/Sources/BraveWallet/Crypto/Onboarding/RestoreWalletView.swift:274]: Desktop has another checkbox
So there are three outcomes:
12 words
24 words
24 words + legacy

image

Comment (supermassive) [reply ios/brave-ios/Sources/BraveWallet/Crypto/Onboarding/RestoreWalletView.swift:274]: It's actually a bigger problem as all mnemonics with 24 words are considered "legacy" now for ios. Which means restoring mnemonic on another wallet app will produce different set of ETH addresses

Comment (StephenHeaps) [reply ios/brave-ios/Sources/BraveWallet/Crypto/Onboarding/RestoreWalletView.swift:274]: @nuo-xu can you take a look at this tomorrow? Looks like toggle was [dropped when v2 UI was integrated](https://github.com/brave/brave-ios/pull/7746), previous would [show toggle when 24-words was selected](https://github.com/brave/brave-ios/blob/e72f2d7bf5f08bead1795e3e2f03016c52a42943/Sources/BraveWallet/Crypto/Onboarding/RestoreWalletView.swift#L105). `isBraveLegacyWallet` state value there, but no toggle for it and v2 UI is instead [using word count](https://github.com/brave/brave-core/blob/9f204d486f58a098f9687464150e24df0bb2160e/ios/brave-ios/Sources/BraveWallet/Crypto/Onboarding/RestoreWalletView.swift#L33-L35).

Comment (nuo-xu) [reply ios/brave-ios/Sources/BraveWallet/Crypto/Onboarding/RestoreWalletView.swift:274]: [issue ](https://github.com/brave/brave-browser/issues/38123) created. will fix asap.

File: components/brave_wallet/browser/keyring_service.cc [hunk 5]
```diff
@@ -2018,151 +1968,182 @@ void KeyringService::MaybeMigratePBKDF2Iterations(const std::string& password) {
...
-
-void KeyringService::ResetAutoLockTimer() {
-  if (auto_lock_timer_->IsRunning()) {
-    auto_lock_timer_->Reset();
-  } else {
-    size_t auto_lock_minutes =
-        (size_t)profile_prefs_->GetInteger(kBraveWalletAutoLockMinutes);
-    auto_lock_timer_->Start(FROM_HERE, base::Minutes(auto_lock_minutes), this,
-                            &KeyringService::OnAutoLockFired);
+void KeyringService::MaybeMigrateToWalletMnemonic(const std::string& password) {
```
Comment (kdenhartog) [inline components/brave_wallet/browser/keyring_service.cc:1971]: We had originally talked about a way to get rid of some of these migrations, but @supermassive thought this approach would be a bit easier to deal with. this approach LGTM

**Review by kdenhartog — APPROVED** [review]
no concerns from me with this

**Pull request merged**